src/System.Management.Automation/engine/parser/TypeInferenceVisitor.cs

```diff
@@ -2930,7 +2930,7 @@ private sealed class VariableAssignmentVisitor : AstVisitor2
 
             private void SetLastAssignment(Ast ast, bool enumerate = false, bool redirectionAssignment = false)
             {
-                if (LastAssignmentOffset < ast.Extent.StartOffset)
+                if (LastAssignmentOffset < ast.Extent.StartOffset && !VariableTarget.Extent.IsWithin(ast.Extent))
```

Comment:
Have we the same issue in next method SetLastAssignmentType?

Reply:
My initial thought was no because it's only used for variable redirection and the OutVar parameters in commands so there was no fear of infinite loops. However, for the sake of correctness I've also fixed that so it excludes variables assigned like that when inferring a variable inside the commandAst as you wouldn't be able to use it there anyway.

Reply:
Since it is not related to infinite loop please move the commit to new PR.